### PR TITLE
Asthma open source calendar re-host on new scheduler

### PR DESCRIPTION
**Pull request by ghost**
- Implement an activity indicator
- Make bracket spacing consistent
- Add observer for activity completion notification
- Update fetch requests for new scheduler

### CODE REVIEW

File: Asthma/TabBarControllers/Dashboard/APHCalendarDataModel.m [hunk 1]
```diff
@@ -194,128 +228,85 @@ -(NSDictionary *)userCompliedWithDailyScheduledTasks:(APHCalendarTaskType)task i
...
-                }else{
-                    [complianceDictionary setObject:@"1" forKey:[NSString stringWithFormat:@"%i", day]];
+                }else if ([dictionary[kDaytimeSickKey] isEqualToNumber: @0]){
+                    [complianceDictionary setObject:@"1" forKey:[NSString stringWithFormat:@"%i", day]];//1 = compliance/green color
                 }
             }
         }
+        
+        dispatch_async(dispatch_get_main_queue(), ^{
+            [[NSNotificationCenter defaultCenter]postNotificationName:calendarDataSourceDidUpdateComplianceDictionaryNotification object:complianceDictionary];
```
Comment (YuanZhu-apple) [inline Asthma/TabBarControllers/Dashboard/APHCalendarDataModel.m]: Add a space between `]` and `post...`

File: Asthma/TabBarControllers/Dashboard/APHCalendarDataModel.m [hunk 2]
```diff
@@ -194,128 +228,85 @@ -(NSDictionary *)userCompliedWithDailyScheduledTasks:(APHCalendarTaskType)task i
...
-    return complianceDictionary;
 }
 
-+ (NSArray *)allScheduledTasksForDateRange: (APCDateRange*) dateRange completed: (NSNumber*) completed inContext: (NSManagedObjectContext*) context
-{
-    NSFetchRequest * request = [APCScheduledTask request];
-    request.shouldRefreshRefetchedObjects = YES;
-    NSPredicate * datePredicate = [NSPredicate predicateWithFormat:@"(startOn >= %@) AND (endOn <= %@)", dateRange.startDate, dateRange.endDate];
+#pragma mark Helpers
+-(BOOL)isValidDateForDay:(NSInteger)DD month:(NSInteger)MM year:(NSInteger)YYYY{
```
Comment (YuanZhu-apple) [inline Asthma/TabBarControllers/Dashboard/APHCalendarDataModel.m]: Make this a shared utility method, I have seen this method twice.

screen shot 2015-06-22 at 1 48 40 pm

P.S. No need to keep this method if the "counting days while loop" is removed.

File: Asthma/TabBarControllers/Dashboard/APHCalendarDataModel.m [hunk 3]
```diff
@@ -194,128 +228,85 @@ -(NSDictionary *)userCompliedWithDailyScheduledTasks:(APHCalendarTaskType)task i
...
                 NSDictionary * dictionary = resultSummary ? [NSDictionary dictionaryWithJSONString:resultSummary] : nil;
                 NSDateComponents *weekdayComponents = [gregorian components:(NSCalendarUnitDay) fromDate:startOn];
                 int day = (int)[weekdayComponents day];
-                if (dictionary[kDaytimeSickKey]) {//had symptoms, red == 0
+                if ([dictionary[kDaytimeSickKey] isEqualToNumber: @1]) {//had symptoms
                     [complianceDictionary setObject:@"0" forKey:[NSString stringWithFormat:@"%i", day]];
-                }else{
-                    [complianceDictionary setObject:@"1" forKey:[NSString stringWithFormat:@"%i", day]];
+                }else if ([dictionary[kDaytimeSickKey] isEqualToNumber: @0]){
+                    [complianceDictionary setObject:@"1" forKey:[NSString stringWithFormat:@"%i", day]];//1 = compliance/green color
```
Comment (YuanZhu-apple) [inline Asthma/TabBarControllers/Dashboard/APHCalendarDataModel.m]: Make predefined keys for @"1" and @"0"?

File: Asthma/TabBarControllers/Dashboard/APHCalendarDataModel.m [hunk 4]
```diff
@@ -194,128 +228,85 @@ -(NSDictionary *)userCompliedWithDailyScheduledTasks:(APHCalendarTaskType)task i
...
+            NSDate *startOn = scheduledTask.startOn;
             if ([startOn earlierDate:[NSDate new]] == startOn) {//no need to parse scheduled tasks after today
-                NSString * resultSummary = task.lastResult.resultSummary;
+                NSString * resultSummary = scheduledTask.lastResult.resultSummary;
                 NSDictionary * dictionary = resultSummary ? [NSDictionary dictionaryWithJSONString:resultSummary] : nil;
                 NSDateComponents *weekdayComponents = [gregorian components:(NSCalendarUnitDay) fromDate:startOn];
                 int day = (int)[weekdayComponents day];
-                if (dictionary[kDaytimeSickKey]) {//had symptoms, red == 0
+                if ([dictionary[kDaytimeSickKey] isEqualToNumber: @1]) {//had symptoms
                     [complianceDictionary setObject:@"0" forKey:[NSString stringWithFormat:@"%i", day]];
```
Comment (YuanZhu-apple) [inline Asthma/TabBarControllers/Dashboard/APHCalendarDataModel.m]: Use a inline static method to convert day to string?

File: Asthma/TabBarControllers/Dashboard/APHCalendarDataModel.m [hunk 5]
```diff
@@ -137,10 +165,12 @@ -(NSDictionary *)userCompliedWithDailyScheduledTasks:(APHCalendarTaskType)task i
                 dateToExamine = [dateToExamine dateByAddingDays:1];
                 monthComponents = [gregorian components:(NSCalendarUnitMonth) fromDate:dateToExamine];
             }
-            
         }
         
-        return complianceDictionary;
+        
+        dispatch_async(dispatch_get_main_queue(), ^{
+            [[NSNotificationCenter defaultCenter]postNotificationName:calendarDataSourceDidUpdateComplianceDictionaryNotification object:complianceDictionary];
```
Comment (YuanZhu-apple) [inline Asthma/TabBarControllers/Dashboard/APHCalendarDataModel.m]: Need a space between `]` and `post...`

File: Asthma/TabBarControllers/Dashboard/APHCalendarDataModel.m [hunk 6]
```diff
@@ -38,45 +38,74 @@
...
+                         [complianceDictionary setObject:@"" forKey:[NSString stringWithFormat:@"%i", (int)components.day]];
+                     }else if ((float)completed / (float)required > kParticipationTrophyThreshold){
+                         [complianceDictionary setObject:@"1" forKey:[NSString stringWithFormat:@"%i", (int)components.day]];
+                     }else{
+                         [complianceDictionary setObject:@"0" forKey:[NSString stringWithFormat:@"%i", (int)components.day]];
+                     }
+                 }
+                 
+                 dispatch_async(dispatch_get_main_queue(), ^{
+                     [[NSNotificationCenter defaultCenter]postNotificationName:calendarDataSourceDidUpdateComplianceDictionaryNotification object:complianceDictionary];
```
Comment (YuanZhu-apple) [inline Asthma/TabBarControllers/Dashboard/APHCalendarDataModel.m]: Need a space between `]` and `post...`

File: Asthma/TabBarControllers/Dashboard/APHCalendarDataModel.m [hunk 7]
```diff
@@ -159,25 +189,29 @@ -(NSDictionary *)userCompliedWithDailyScheduledTasks:(APHCalendarTaskType)task i
...
-                    [complianceDictionary setObject:@"1" forKey:[NSString stringWithFormat:@"%i", day]];
+                }else if ([dictionary[kNighttimeSickKey] isEqualToNumber: @0]){
+                    [complianceDictionary setObject:@"1" forKey:[NSString stringWithFormat:@"%i", day]];//1 = compliance/green color
                 }
             }
         }
-        return complianceDictionary;
+        
+        dispatch_async(dispatch_get_main_queue(), ^{
+            [[NSNotificationCenter defaultCenter]postNotificationName:calendarDataSourceDidUpdateComplianceDictionaryNotification object:complianceDictionary];
```
Comment (YuanZhu-apple) [inline Asthma/TabBarControllers/Dashboard/APHCalendarDataModel.m]: Need a space between `]` and `post...`

File: Asthma/TabBarControllers/Dashboard/APHCalendarDataModel.m [hunk 8]
```diff
@@ -38,45 +38,74 @@
...
+        
+        while ([self isValidDateForDay:endDay month:month year:year]) {
+            ++endDay;
         }
         
-        return complianceDictionary;
+        [comps setDay:endDay -1];
+        [comps setMonth:month];
+        [comps setYear:year];
+        NSDate *endDate = [gregorian dateFromComponents:comps];
```
Comment (YuanZhu-apple) [inline Asthma/TabBarControllers/Dashboard/APHCalendarDataModel.m:74]: Some better ways to find the last day of a month (avoid a while loop):
http://stackoverflow.com/questions/2772348/getting-the-last-day-of-a-month

File: Asthma/TabBarControllers/Dashboard/APHCalendarDataModel.m [hunk 9]
```diff
@@ -194,128 +228,85 @@ -(NSDictionary *)userCompliedWithDailyScheduledTasks:(APHCalendarTaskType)task i
...
-        if ([scheduledTask.dateRange compare:dateRange] != kAPCDateRangeComparisonOutOfRange) {
-            [filteredArray addObject:scheduledTask];
-        }
-    }
-    return filteredArray.count ? filteredArray : nil;
+    return dailyScheduledTasks;
 }
 
-+ (NSArray *)allScheduledWeeklyPromptTasksForDateRange: (APCDateRange*) dateRange completed: (NSNumber*) completed inContext: (NSManagedObjectContext*) context
+- (NSArray *)weeklySurveyScheduledTasksForDateRange: (APCDateRange *)dateRange
```
Comment (YuanZhu-apple) [inline Asthma/TabBarControllers/Dashboard/APHCalendarDataModel.m]: Can this be consolidated with `dailySurveyScheduledTasksForDateRange:`?
I noticed the only difference is `task.taskID`.

File: Asthma/TabBarControllers/Dashboard/APHCalendarCollectionViewController.m [hunk 10]
```diff
@@ -97,23 +102,39 @@ - (void)viewDidLoad {
...
+{
+    [self performSelector:@selector(dismiss) withObject:self afterDelay:0.5];
+    self.compliance = (NSDictionary *)[notification object];
     [self.collectionView reloadData];
 }
 
--(void)setCalendarAttributes{
+-(void)dealloc
+{
+    [[NSNotificationCenter defaultCenter]removeObserver:self];
```
Comment (YuanZhu-apple) [inline Asthma/TabBarControllers/Dashboard/APHCalendarCollectionViewController.m]: Need a space.

File: Asthma/TabBarControllers/Dashboard/APHCalendarCollectionViewController.m [hunk 11]
```diff
@@ -85,11 +86,15 @@ - (void)viewDidLoad {
     if (self.taskType == kAPHCalendarTaskTypeFreeNights) {
         self.title = NSLocalizedString(@"Undisturbed Nights", nil);
     }
-    
+
+    [[NSNotificationCenter defaultCenter]addObserver:self selector:@selector(reloadCollectionView:) name:calendarDataSourceDidUpdateComplianceDictionaryNotification object:nil];
```
Comment (YuanZhu-apple) [inline Asthma/TabBarControllers/Dashboard/APHCalendarCollectionViewController.m]: Space before `addObserver`

**Comment by ghost** [comment]
@YuanZhu-apple  Comments addressed. Thanks.

File: Asthma/TabBarControllers/Dashboard/APHCalendarDataModel.m [hunk 12]
```diff
@@ -120,16 +153,16 @@ -(NSDictionary *)userCompliedWithDailyScheduledTasks:(APHCalendarTaskType)task i
...
-                NSString * resultSummary = task.lastResult.resultSummary;
+                NSString * resultSummary = scheduledTask.lastResult.resultSummary;
                 NSDictionary * dictionary = resultSummary ? [NSDictionary dictionaryWithJSONString:resultSummary] : nil;
                 
                 //check we're still in the current month
                 if ((int)monthComponents.month == (int)month) {
                     NSString *keyString = [kDaysMissedKey stringByAppendingFormat:@"%i", (int)weekday];
                     if (dictionary[keyString]) {//missed work, red == 0
-                        [complianceDictionary setObject:@"0" forKey:[NSString stringWithFormat:@"%i", day]];
+                        [complianceDictionary setObject:kMissedWorkBoolean forKey:[NSString stringWithFormat:@"%i", day]];
```
Comment (YuanZhu-apple) [inline Asthma/TabBarControllers/Dashboard/APHCalendarDataModel.m]: Use a inline static method to convert a day to string?

File: Asthma/TabBarControllers/Dashboard/APHCalendarDataModel.m [hunk 13]
```diff
@@ -35,48 +35,82 @@
...
+                     NSArray *tasks = [dates objectForKey:date];
+                     for (APCTaskGroup *taskGroup in tasks) {
+                         required += [taskGroup requiredRemainingTasks].count + [taskGroup requiredCompletedTasks].count;
+                         completed += [taskGroup requiredCompletedTasks].count;
+                     }
+                     
+                     NSDateComponents* components = [gregorian components:NSCalendarUnitYear|NSCalendarUnitMonth|NSCalendarUnitDay fromDate:date];
+                     
+                     if ([date compare:[NSDate new]] == NSOrderedDescending) {
+                         [complianceDictionary setObject:@"" forKey:[NSString stringWithFormat:@"%i", (int)components.day]];
```
Comment (YuanZhu-apple) [inline Asthma/TabBarControllers/Dashboard/APHCalendarDataModel.m]: Use a shared utility method to convert day to string.
An what is `setObject:@""`?

**Comment by ghost** [comment]
@YuanZhu-apple  Latest commit requires a couple of updates to AppCore. Will send a PR shortly.

File: Asthma/TasksAndSteps/APHDailyTaskViewController.m [hunk 14]
```diff
@@ -80,10 +80,11 @@ - (NSString*) createResultSummary
     //day symptoms
     {
         ORKBooleanQuestionResult *result = (ORKBooleanQuestionResult *)[self answerForSurveyStepIdentifier:kDaySymptomsIdentifier];
-            if ([result booleanAnswer] == [NSNumber numberWithBool:1]) {
+        if ([result booleanAnswer] == [NSNumber numberWithBool:1]) {
             dictionary[kDaytimeSickKey] = @1;
+        }else if([result booleanAnswer] == [NSNumber numberWithBool:0]){
+            dictionary[kDaytimeSickKey] = @0;
         }
```
Comment (YuanZhu-apple) [inline Asthma/TasksAndSteps/APHDailyTaskViewController.m:86]: Why not just do:

```
 dictionary[kDaytimeSickKey] = [result booleanAnswer];
```

File: Asthma/TasksAndSteps/APHDailyTaskViewController.m [hunk 15]
```diff
@@ -91,6 +92,8 @@ - (NSString*) createResultSummary
         ORKBooleanQuestionResult *result = (ORKBooleanQuestionResult *)[self answerForSurveyStepIdentifier:kNightSymptomsIdentifier];
         if ([result booleanAnswer] == [NSNumber numberWithBool:1]) {
             dictionary[kNighttimeSickKey] = @1;
+        }else if([result booleanAnswer] == [NSNumber numberWithBool:0]){
+            dictionary[kNighttimeSickKey] = @0;
         }
```
Comment (YuanZhu-apple) [inline Asthma/TasksAndSteps/APHDailyTaskViewController.m:95]: ?

```
 dictionary[kDaytimeSickKey] = [result booleanAnswer];
```

File: Asthma/TasksAndSteps/APHDailyTaskViewController.m [hunk 16]
```diff
@@ -98,13 +101,13 @@ - (NSString*) createResultSummary
...
-    if ([result numericAnswer]) {
-        NSNumber * peakFlow = [result numericAnswer] ? [result numericAnswer] : @0;
-        if (peakFlow) {
-            dictionary[kPeakFlowKey] = peakFlow;
+        if ([result numericAnswer]) {
+            NSNumber * peakFlow = [result numericAnswer] ? [result numericAnswer] : @0;
+            if (peakFlow) {
+                dictionary[kPeakFlowKey] = peakFlow;
+            }
         }
```
Comment (YuanZhu-apple) [inline Asthma/TasksAndSteps/APHDailyTaskViewController.m]: Are both `if` condition necessary?  
And personally I prefer `@(NO)` over `@0`, if @0 stands for NO.

File: Asthma/TabBarControllers/Dashboard/APHCalendarDataModel.m [hunk 17]
```diff
@@ -194,128 +226,61 @@ -(NSDictionary *)userCompliedWithDailyScheduledTasks:(APHCalendarTaskType)task i
...
                 NSDictionary * dictionary = resultSummary ? [NSDictionary dictionaryWithJSONString:resultSummary] : nil;
-                NSDateComponents *weekdayComponents = [gregorian components:(NSCalendarUnitDay) fromDate:startOn];
-                int day = (int)[weekdayComponents day];
-                if (dictionary[kDaytimeSickKey]) {//had symptoms, red == 0
-                    [complianceDictionary setObject:@"0" forKey:[NSString stringWithFormat:@"%i", day]];
-                }else{
-                    [complianceDictionary setObject:@"1" forKey:[NSString stringWithFormat:@"%i", day]];
+                if ([dictionary[kDaytimeSickKey] isEqualToNumber: @1]) {//had symptoms
+                    [complianceDictionary setObject:kHadSymptomsBoolean forKey:[self dayFromDate:startOn]];
+                }else if ([dictionary[kDaytimeSickKey] isEqualToNumber: @0]){
```
Comment (YuanZhu-apple) [inline Asthma/TabBarControllers/Dashboard/APHCalendarDataModel.m]: Personally, I prefer @(NO) over @0.

**Comment by ghost** [comment]
@YuanZhu-apple  done

File: Asthma/TabBarControllers/Dashboard/Objects/APHAsthmaBadgesObject.m [hunk 18]
```diff
@@ -189,7 +225,7 @@ - (void) calculateAsthmaFreeDaysValue
         [dailyPromptTasks enumerateObjectsUsingBlock:^(APCScheduledTask * obj, NSUInteger __unused idx, BOOL __unused *stop) {
             NSString * resultSummary = obj.lastResult.resultSummary;
             NSDictionary * dictionary = resultSummary ? [NSDictionary dictionaryWithJSONString:resultSummary] : nil;
-            if (!dictionary[kDaytimeSickKey] && ![dictionary[kDaytimeSickKey] boolValue]) {
+            if ([dictionary[kDaytimeSickKey] isEqualToNumber:@0]) {
```
Comment (YuanZhu-apple) [inline Asthma/TabBarControllers/Dashboard/Objects/APHAsthmaBadgesObject.m]: `@(NO)`

File: Asthma/TabBarControllers/Dashboard/Objects/APHAsthmaBadgesObject.m [hunk 19]
```diff
@@ -211,7 +247,7 @@ - (void) calculateUndisturbedNightsValue
         [dailyPromptTasks enumerateObjectsUsingBlock:^(APCScheduledTask * obj, NSUInteger __unused idx, BOOL __unused *stop) {
             NSString * resultSummary = obj.lastResult.resultSummary;
             NSDictionary * dictionary = resultSummary ? [NSDictionary dictionaryWithJSONString:resultSummary] : nil;
-            if (!dictionary[kNighttimeSickKey] && ![dictionary[kNighttimeSickKey] boolValue]) {
+            if ([dictionary[kNighttimeSickKey] isEqualToNumber:@0]) {
```
Comment (YuanZhu-apple) [inline Asthma/TabBarControllers/Dashboard/Objects/APHAsthmaBadgesObject.m]: `@(NO)`

File: Asthma/TabBarControllers/Dashboard/Objects/APHAsthmaBadgesObject.m [hunk 20]
```diff
@@ -136,13 +138,47 @@ - (NSArray *)completedWeeklyScheduledTasks
...
+    //get states for the dates we're interested in -4 and today
+    NSDate *dateToQuery = [[NSDate new]dateByAddingDays:-4];
+    while ([dateToQuery compare:[NSDate new]] == NSOrderedAscending || [dateToQuery compare:[NSDate new]] == NSOrderedSame) {
+        states = [dateState activitiesStateForDate:dateToQuery];
+        dailySurveyState = [states objectForKey:kDailySurveyTaskID];
+        weeklySurveyState = [states objectForKey:kWeeklySurveyTaskID];
+        
+        //iterate over times in state
+        for (NSDate *time in dailySurveyState) {
+            if ([[dailySurveyState objectForKey:time] isEqualToNumber:@1]) {
```
Comment (YuanZhu-apple) [inline Asthma/TabBarControllers/Dashboard/Objects/APHAsthmaBadgesObject.m]: `@(YES)`

File: Asthma/TabBarControllers/Dashboard/Objects/APHAsthmaBadgesObject.m [hunk 21]
```diff
@@ -136,13 +138,47 @@ - (NSArray *)completedWeeklyScheduledTasks
...
+                completedActivities ++;
+            }
+            totalActivities++;
+        }
+        
+        for (NSDate *time in weeklySurveyState) {
+            if ([[weeklySurveyState objectForKey:time] isEqualToNumber:@1]) {
```
Comment (YuanZhu-apple) [inline Asthma/TabBarControllers/Dashboard/Objects/APHAsthmaBadgesObject.m]: `@(YES)`

**Comment by ghost** [comment]
@YuanZhu-apple Updated

**Pull request merged**